### PR TITLE
Use the expected method depending on the target being a file or folder

### PR DESCRIPTION
**Pull request by jvillafanez**
## Description
Use the expected method to delete the target file or folder

## Related Issue

https://github.com/owncloud/core/issues/26278
Might affect to https://github.com/owncloud/enterprise/issues/1720

## Motivation and Context

Storages are expected to delete files with `unlink` and folders with `rmdir`. Using `unlink` for folders can have unexpected consecuences (see linked core issue)

## How Has This Been Tested?

Manual tested using SMB storage with session credentials. The account has read-only access to the SMB storage.
NOTE: It's not intended that the file will be deleted. As stated in https://github.com/owncloud/core/issues/26278, the folder won't be deleted and an error message will be shown, but the log will be different.
```
{"reqId":"foIHTwhq52ee\/72NdaBU","remoteAddr":"10.0.2.4","app":"webdav","message":"Exception: {\"Message\":\"HTTP\\\/1.1 403 Forbidden\",\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\Forbidden\",\"Code\":0,\"Trace\":\"#0 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Tree.php(179): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Directory->delete()\\n#1 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(287): Sabre\\\\DAV\\\\Tree->delete('windows\\\/dummy\\\/B...')\\n#2 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpDelete(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#3 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#4 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:DELETE', Array)\\n#5 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#6 \\\/opt\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(58): Sabre\\\\DAV\\\\Server->exec()\\n#7 \\\/opt\\\/owncloud\\\/remote.php(164): require_once('\\\/opt\\\/owncloud\\\/a...')\\n#8 {main}\",\"File\":\"\\\/opt\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Directory.php\",\"Line\":301,\"User\":\"user1\"}","level":0,"time":"2017-01-25T16:19:03+00:00","method":"DELETE","url":"\/remote.php\/webdav\/windows\/dummy\/B2","user":"user1"}
```
notice the forbidden exception instead of the invalid type exception.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

Backports required.

@PVince81 @DeepDiver1975

### CODE REVIEW

**Comment by PVince81** [comment]
Good workaround I guess. This `rmdir` vs `unlink` stuff is so messy, it's not the first time: https://github.com/owncloud/core/issues/8630

Let's first see if it solves the original issue.

File: apps/files_trashbin/lib/Trashbin.php [hunk 1]
```diff
@@ -253,7 +253,11 @@ public static function move2trash($file_path) {
 		}
 
 		if ($sourceStorage->file_exists($sourceInternalPath)) { // failed to delete the original file, abort
-			$sourceStorage->unlink($sourceInternalPath);
+			if ($sourceStorage->is_dir($sourceInternalPath)) {
+				$sourceStorage->rmdir($sourceInternalPath);
+			} else {
+				$sourceStorage->unlink($sourceInternalPath);
```
Comment (PVince81) [inline apps/files_trashbin/lib/Trashbin.php]: This code is only executed when the original file could not be deleted apparently.
If that is correct you'd also need to change line 250 above.

Comment (jvillafanez) [inline apps/files_trashbin/lib/Trashbin.php]: The file is removed first from the trash storage, probably to make prevent any possible issue if the target file already exists in the trashbin. This removal happens in the trash storage which is expected to support the unlink operation to delete both files and directories.

I'd rather keep things like this in this PR and make your changes in another PR to fix that behaviour.

**Comment by jvillafanez** [comment]
I'd merge this anyway just to stay safe.

**Comment by jvillafanez** [comment]
Any backports to be made?

**Comment by PVince81** [comment]
At this point I'd say no as it is not clear how to reproduce this accurately.
I see this PR mostly as hardening.
